File: 0394_med_hot_decode-string/solution.java
import java.util.Stack;

class Solution {
    public String decodeString(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c != ']') {
                stack.push(c);
            } else {
                // pop and decode
                StringBuilder token = new StringBuilder();
                while (stack.peek() != '[') {
                    token.insert(0, stack.pop());
                }
                stack.pop();  // remove '['

                StringBuilder numStr = new StringBuilder();
                while (!stack.isEmpty() && Character.isDigit(stack.peek())) {
                    numStr.insert(0, stack.pop());
                }

                int repeatCount = Integer.parseInt(numStr.toString());
                String repeatedToken = token.toString().repeat(repeatCount);
                
                for (char ch : repeatedToken.toCharArray()) {
                    stack.push(ch);
                }
            }
        }

        StringBuilder result = new StringBuilder();
        for (char ch : stack) {
            result.append(ch);
        }

        return result.toString();
    }
}

File: 0394_med_hot_decode-string/solution.py


'''
Example 1:

Input: s = "3[a]2[bc]"
Output: "aaabcbc"
Example 2:

Input: s = "3[a2[c]]"
Output: "accaccacc"
Example 3:

Input: s = "2[abc]3[cd]ef"
Output: "abcabccdcdcdef"
'''
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []

        for c in s:
            if c != ']':
                stack.append(c)
            else:
                # pop and decode
                token = ""
                while stack[-1] != '[':
                    token = stack.pop() + token
                stack.pop()

                num_str = ""

                while stack[-1].isdigit():
                    num_str = stack.pop() + num_str
                stack.extend(token * int(num_str))
                
        return ''.join(stack)
